### ISSUE TITLE
Game will not close using window close button (Unique Situation)

### ISSUE THREAD

**Issue by KinsonDigital** [MonoGame/MonoGame#6432]
I am having an issue with the window/game freezing when clicking the window's X close button.

Hold onto your hats. I am going to try to explain this the best that I can and I am sure that the issue has to do with the way I am doing this and using **MonoGame**.  I just don't know how to fix it.

I am creating a game engine that has a **plugin type architecture**. The whole idea is that my engine will be able to load implementations of various parts of my engine, by loading a DLL library at runtime that contains those implementations.

Doing this will let me (or other developers) actually take only certain parts of my engine and swap out the implementation with another one as long as they adhere to the interfaces provided. This shared library that contains all the interfaces is in its own library.  The implementation library is another separate project.

This implementation library would be the **plugin** for the underlying core of the engine.

So the way it would work would be this.  Developer John Doe would create a C# windows project running .NET Framework 4.5.  This project is the **game** and would have a reference to my engine library for use.  John Doe would then create a C# library project using .NET Framework 4.5 under the same solution.  For this library project he would reference my **engine core** library that contains all of the interfaces that he has to adhere to for my game engine to use the new implementation library he is creating.

This implementation library is loaded during runtime by my game engine.  Throughout my engine I use a static class to pull a particular object implementation from the assembly that was loaded and inject it where its needed.

---

Example Of Usage:

I create the windows app, reference the game engine.  Create another library project called **MyImplementationA** and I reference my engine core library.  In the implementation library I just created I would create a class called **TextureA** that implements the **ITexture** interface that comes from the core library. I write my **implementation** for **TextureA**.

Now I create another implementation library project called **MyImplementationB** and reference my engine core library just like above.  But this time I create a class called **TextureB** that implements the **ITexture** interface and write its implementation.

Now I have 2 assemblies.  **MyImplementationA.dll** implements **TextureA** and **MyImplementationB.dll** implements **TextureB**.  Now at this point I can just simply load either library up and I get that particular functionality.  

Doing this means that if **DevA** wants to use **MonoGame** to do some of the work for the engine under the hood, 
and **DevB** wants to use **SDL**, it will all operate the same.  Or they can use something else if they wanted.

Now everything I explained above works like a charm.  Its pretty awesome and I am very happy with the results.

But.....my implementation that I am using for testing my game engine is MonoGame.  This means in the implementation library it uses the **Game** class.  I am only assuming that this is the cause because I technically am loading the **MonoGame** **Game** during runtime from another assembly.  Could this be the issue?  I know that it must be obviously running in the same process as my game engine, because my engine is the one loading the assembly so that means that **MonoGame** under the hood would be running in the same process as my engine.

One thing I tried to do is have my engine get access to the **Game.Exit()** method so that way I could try to exit the game and close the window in code.  This does not work.  It just throws a **System.NullReferenceException** at the line of code where **Game.Run()** is used.  This tells me that at this point **Game** has been disposed of, garbage collected, and doesn't exist anymore.  So I guess it is **"Exiting"**, but not closing the window.  And with that I think the reason the window doesn't close is because it isn't **"attached"** to **MonoGame** anymore.

I know this situation is unique and there are probably a lot of unknowns, but at this point I don't know what to do about the issue and really don't want this to prevent me from using **MonoGame** or any other libraries.

If you need more information or have some questions for me to help you understand, don't hesitate to ask.

Any help would greatly be appreciated!!

---

#### What operating system are you using:
- Windows 10

#### What version of MonoGame does the bug occur on:
- MonoGame 3.6

#### What MonoGame platform are you using:
- DesktopGL

**Comment by KinsonDigital** [comment]
Also, here is an example project showing the issue.

[Issue GitHub Repo](https://github.com/KinsonDigital/MonoFreezeOnCloseIssue)

The way this is setup is the same in my original project.  Of course there is much more source code in the original project but this is the bare bones of what I am doing to pull off what I want to do.  Also since I can repeat the issue with the project in the link above, I think I have proven that the issue is definitely with me loading MonoGame that is embed in another assembly and run time.

I will keep working on the issue and will keep updating this post as things change.

**Comment by KinsonDigital** [comment]
**Update**

I tried to registering to the **Game.Existing** event and put a break point in the event method.  I ran the application and tried to close the window and the event is not invoked.

**Comment by KinsonDigital** [comment]
**Another Update**

So I went little deeper.  This is what I did this time.  

1. I made 2 separates copies of the **MonoGame** source and put them into separate folders. One copy was for DirectX testing and the other for OpenGL testing.
2. I took the sample projects in the post above and put them into the main MonoGame source for both source locations copies.
3. I properly setup the references to point to the main source framework project.
4. Made sure that all of the binaries from the proper places were pushed to the proper place using build events.
5. Ran my **MainGame** project and closed the window using the window X button. I tried this for both copies and setups.

**Results:**

The DirectX version of worked flawlessly.  No freezing and the window closed.

The OpenGL version did not work.

My original project that I was having issues with this on is an OpenGL project as well. So I have narrowed this down to an OpenGL issue.  I don't know why that makes a difference but I am going to start stepping through the MonoGame source and try to see what the issue is.

**Comment by KinsonDigital** [comment]
@tomspilman 
@cra0zy 

Just was wondering if you guys had a chance to look at this?  Very easy fix and could go into v3.7.1.   :)

**Comment by cra0zy** [comment]
Can you summarize what the issue is and what the solution is in few sentence please.

**Comment by KinsonDigital** [comment]
Sure thing.

**Issue:**

When instantiating a **MonoGame** Game object in a assembly and loading that assembly during run time from a console application for use, clicking the close button to close a **MonoGame** window will not close the window.

This seems to only occur in an OpenGL game.

**Solution:**

Add 2 lines of code after the **Line: 181** in the file **SDLGamePlatform.cs** file in the **SdlRunLoop()** method.

Doing this will properly check for the **ev.Window.EventID** to see if it is the **Sdl.Window.EventId.Close** value. Previously this value was not being checked.

**Code Before:**

```
    if (ev.Window.EventID == Sdl.Window.EventId.Resized || ev.Window.EventID == Sdl.Window.EventId.SizeChanged)
	_view.ClientResize(ev.Window.Data1, ev.Window.Data2);
    else if (ev.Window.EventID == Sdl.Window.EventId.FocusGained)
	IsActive = true;
    else if (ev.Window.EventID == Sdl.Window.EventId.FocusLost)
	IsActive = false;
    else if (ev.Window.EventID == Sdl.Window.EventId.Moved)
    	_view.Moved();
```

**Code After:**
```
    if (ev.Window.EventID == Sdl.Window.EventId.Resized || ev.Window.EventID == Sdl.Window.EventId.SizeChanged)
	_view.ClientResize(ev.Window.Data1, ev.Window.Data2);
    else if (ev.Window.EventID == Sdl.Window.EventId.FocusGained)
	IsActive = true;
    else if (ev.Window.EventID == Sdl.Window.EventId.FocusLost)
	IsActive = false;
    else if (ev.Window.EventID == Sdl.Window.EventId.Moved)
        _view.Moved();
    else if (ev.Window.EventID == Sdl.Window.EventId.Close)
	_view._game.Exit();		
```

For some reason this is not an issue if your instantiating a **Game** object in the same app that the Game object is being used in.  But from another assembly, the window will not close.  I have tested this and applying this fix does indeed close the window in an **OpenGL** MonoGame application when loaded from an assembly.

As mentioned in the above posts, if you guys are interested in testing this out to reproduce the problem, I built a basic setup project to showcase the issue and that link is in the second from the top post.  

If you need anything else, don't hesitate to ask.

Thanks!!

**Comment by cra0zy** [comment]
I can summarize that entire comment in 2 sentences :\.

**Comment by Jjagg** [comment]
@KinsonDigital Any chance you could set up a minimal project that demonstrates the issue on GitHub? That would make us implementing and testing a solution way easier :)

**Comment by KinsonDigital** [comment]
Sure thing.  It is in the second post from the top. 

I will post the link again.  :)  Thanks!!

[Min Project Demonstrating Issue](https://github.com/KinsonDigital/MonoWillNotCloseIssue)

**Comment by Jjagg** [comment]
Woops, sorry I missed that 😅 Thanks!

**Comment by KinsonDigital** [comment]
Np.  Let me know if you would like me to do a pull request.  I would love to if that is the route you want to take.

**Comment by Jjagg** [comment]
SDL fails to raise the `SDL_QUIT` event in this scenario, even if the last window gets the `SDL_CLOSE` event. I checked their source code and it seems simple enough, so I'm not sure what's up. 

Anyway I set up a PR to fix the issue #6489. Thanks for the good repro + suggestion @KinsonDigital :)

**Comment by KinsonDigital** [comment]
Np.  Glad I could help.
